ruby/phase-1.md

```diff
@@ -0,0 +1,434 @@
+companies to rely on managed solutions. This presents an opportunity for Temporal to become Ruby's
+de facto standard for workflow orchestration.
+
+This proposal aims to adhere to the Ruby way of writing code favouring convention over
+configuration, readability, testability and developer's productivity. These are prioritised in order
+to avoid the destiny of the [AWS Flow Framework for Ruby](https://github.com/amazon-archives/aws-flow-ruby)
+that never appealed to the Ruby community.
+
+
+## Disclaimer
+
+This proposal draws substantial inspiration from the [Temporal Ruby SDK](https://github.com/coinbase/temporal-ruby)
+developed at Coinbase by the author of this proposal. This allows us to pick the best parts of the
+former SDK and improve on the parts that are not so good.
+
+The proposal itself follows the [Python SDK proposal](https://github.com/temporalio/proposals/blob/master/python/phase-1.md)
+for the same exact reason.
+
+
+## Timeline
+
+The Ruby SDK will be implemented in four loosely defined phases:
+
+- Phase 1 — initial setup, execution attributes, workflow/activity outer APIs, client and worker
+- Phase 2 — workflow and activity inner APIs, local activities, async activity completion, etc
+- Phase 3 — payload converters, error handling, middleware
+- Phase 4 — Sandbox, replayer, support for testing
+
+
+## Overview
+
+The Ruby SDK will be implemented on top of the [SDK Core](https://github.com/temporalio/sdk-core).
+The exact bridging mechanism will decided later as there are different alternatives.
+
+### Goals
+
+- Quickest path to a working setup
+- Convention over configuration
+- Avoid sacrificing configurability of a complex setup
+- Minimise external dependencies (gems)
+- Local and end-to-end testability of workflows
+
+### Non-goals
+
+- Replicate the interface of other SDKs
+- Backwards compatibility with Coinbase's Ruby SDK
+
+### Tech stack && approach
+
+- Ruby >= 2.7 (currently oldest fully supported version)
+- Packaged as a Ruby gem (and published to [https://rubygems.org])
+- FFI or Rutie bridge into SDK Core (to be decided)
+
+
+## An end-to-end example
+
+Perhaps a bit unconventionally (as far as the proposal review process goes), but we'd like to start
+with an end-to-end example of a simple and fully functional "Hello World" workflow. The intention of
+this example is to demonstrate how everything ties together before diving into each separate moving
+part and exploring it in depth.
+
+```ruby
+require 'temporal'
+require 'temporal/worker'
+
+Temporal.configure do |config|
+  config.url = 'localhost:7233'
+  config.namespace = 'my-namespace'
+  config.task_queue = 'my-task-queue'
+end
+
+class MyActivity < Temporal::Activity
+  def execute
+    puts 'Hello World!'
+  end
+end
+
+class MyWorkflow < Temporal::Workflow
+  def execute
+    MyActivity.execute!
+  end
+end
+
+worker = Temporal::Worker.new
+worker.register_workflow(MyWorkflow)
+worker.register_activity(MyActivity)
+worker.run
```

Comment:
Does this throw in case there’s a fatal error?

Reply:
It will throw in case of a SystemError (such as OOM, Syntax error, Load error, Signal, etc), but all the workflow/activity errors are tightly controlled inside the worker